Support null-pointer without operand in interpreter

mgudemann opened the pull request:
Before the interpreter evaluate function always expected a pointer type to have an operand, but we can have NULL pointers without operand, in the example it was from `cprover_string_content`

Commit:
Support null-pointer without operand in interpreter

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 624d1aa8205c98f65529253ed1fb401e581995fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84141522

tautschnig commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -339,11 +339,19 @@ void interpretert::evaluate(
         return;
       }
-      if(expr.has_operands() && !to_integer(expr.op0(), i))
+      else if(
```
Sorry, not being changed here, but as I'm trying to understand the code: line 334 suggests there is a _type_ `ID_address_of`?

thk123 replied:
That _surely_ must be in error!

mgudemann replied:
I removed this (and one other occurrence) I don't think those can be types

tautschnig commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -339,11 +339,19 @@ void interpretert::evaluate(
       }
-      if(expr.has_operands() && !to_integer(expr.op0(), i))
+      else if(
+        expr.has_operands() && !to_integer(to_constant_expr(expr.op0()), i))
```
What does a pointer-typed constant (mind the test in line 300) with operands even look like?

thk123 replied:
Constant memory offset? `void *x = 0x100`? Don't know if that is _ever_ what you want

mgudemann replied:
@tautschnig this for example 
```
java::@inflight_exception
constant
  * type: pointer
      * width: 64
      * #no_nondet_initialization: 1
      0: empty
  * value: 0000000000000000000000000000000000000000000000000000000000000000
  0: constant
      * type: pointer
          * width: 64
          * #no_nondet_initialization: 1
          0: empty
      * value: NULL
```
effectively this here was the culprit for the `failed to evaluate expression: null`:
```
cprover_string_content::cprover_string_content
constant
  * type: pointer
      * width: 64
      0: unsignedbv
          * width: 16
  * value: NULL
  * #source_location: 
    * file: org/apache/tika/parser/prt/PRTParser.java
    * line: 222
    * function: java::org.apache.tika.parser.prt.PRTParser.extractText:([BZ)Ljava/lang/String;
    * java_bytecode_index: 22
!! failed to evaluate expression: null
```
(both first name then `.pretty` output)

thk123 replied:
That is definitely wrong (!) Could you raise a ticket either on issues here or on Jira to track this down. @LAJW I bet this is why you had some weird edge case handling code for null pointers!

tautschnig commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -339,11 +339,19 @@ void interpretert::evaluate(
+      // check if expression is constant null pointer without operands
+      else if(
+        !expr.has_operands() && !to_integer(to_constant_expr(expr), i) &&
+        i == 0)
```
You might just do `expr.is_zero()`, although `config.ansi_c.NULL_is_zero` should actually be considered.

mgudemann replied:
made it `i.is_zero()` I am not sure that java correctly sets the `NULL_is_zero` in any case

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8b3bfaac75dfbe0f724d72ff0104874640bc5b4e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84148025

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 56ef1d4ec100c2b82e97849774c129a137b2271a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84153742

thk123 reviewed (commented):
I don't know how feasible this is, but I have no idea what this change does so it'd be great to ad a (unit?) test.

Commit:
Remove ID_address_of check for types as it can only be exprt

mgudemann commented:
@thk123 this fixes an error message from the interpreter. It should be possible to call the `evaluate` method f the interpreter and to check the result.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 55a1dfe80dac5c54912a4122fc2a7d4380fecc22).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84304603

Commit:
Add unit test for interpreter evaluation

smowton commented:
I don't have a strong opinion which one we want, but surely there should be one correct form of a null pointer constant, and the interpreter should understand that one form?

@tautschnig the "operand" form is I believe inspired by the constant object pointers produced by prop_convt::get, which will make something like 0x100 (operand &obj_xyz), thus providing both the address and the meaning of that address. The null with operands is giving the numerical representation (0) and the symbolic one (null).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: bce73d488a31055d92c4c94e6672ae58d912daa6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84403064
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

thk123 commented:
@smowton - thanks for this that makes sense. But `null_pointer_exprt` returns the no-operand version - so it should be updated to return this op'd version if this is really is the "correct" format for a pointer. This should also be documented (perhaps in `pointer_typet`/`null_pointer_exprt`/`address_of_exprt`)

thk123 commented on src/goto-programs/interpreter_class.h:
```diff
@@ -306,6 +306,14 @@ class interpretert:public messaget
   void initialize(bool init);
   void show_state();
+
+  friend mp_vectort
```
:no_entry_sign: This is required for the unit test right? I think it is cleaner to declare `friend class interpreter_testt` and then put this into the `unit` folder 0 as it stands you might as well make `evaluate` public no?

mgudemann replied:
will go with the friend class, that is hopefully less controversial

thk123 commented on unit/interpreter/interpreter.cpp:
```diff
@@ -0,0 +1,34 @@
+  interpretert::mp_vectort mp_vector =
+    interpreter_evaluate(interpreter, constant_expr);
+
+  REQUIRE(mp_vector.size() == 1);
```
:pick: You can use
```C++
REQUIRE_THAT(mp_vector, Catch::Matchers::Vector::Equals<mp_integer>{{0}});
```
To check that the array is exactly `{0}`

thk123 reviewed (commented):
Thanks for the test!

:bulb: can we get one for the other kind of null pointer too?
:no_entry_sign: avoid putting code only required for test in the main `src`

smowton commented:
It might be better to say there is a correct format in a particular place. The interpreter operates over traces, which have been produced by prop_convt::get'ing all the symbols in a particular equation. My guess is it's behaving differently for constants given in the input equation vs. symbols that have been solved for a zero / null value.

mgudemann commented:
@thk123 addressed your comments

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: d4f9fb058abecc9041282a8e6df8cc80f89f0066).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84421606
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

majakusber commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -339,7 +339,16 @@ void interpretert::evaluate(
         return;
       }
-      if(expr.has_operands() && !to_integer(expr.op0(), i))
+      else if(
```
Since all of the `if`/`else if` return, you can make them all `if`s.

majakusber commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -339,7 +339,16 @@ void interpretert::evaluate(
       }
-      if(expr.has_operands() && !to_integer(expr.op0(), i))
+      else if(
+        expr.has_operands() && !to_integer(to_constant_expr(expr.op0()), i))
```
Does `expr.op0()` always need to be a constant in this case?

mgudemann replied:
not necessarily, will change this back

majakusber commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -339,7 +339,16 @@ void interpretert::evaluate(
+      }
+      // check if expression is constant null pointer without operands
+      else if(
+        !expr.has_operands() && !to_integer(to_constant_expr(expr), i) &&
```
Does `expr` always need to be a constant in this case?

mgudemann replied:
yes, it is tested for having `id == ID_constant`

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d22f26aa1f9491651895281a6a33fb3cd4ad007b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84449831

romainbrenguier committed:
Correct input initialization in builtin function

Result was moved twice instead of initializing `input` field with the
given `input`.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 7bc468d891c058667808008a60627b244f05a084).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84545908

mgudemann commented:
based on #2935

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f7e9715588aba6be08a5b32733a42efc2123054a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84558263
